poly/joint.go

```diff
@@ -0,0 +1,284 @@
+package poly
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"github.com/dedis/crypto/abstract"
+	"github.com/dedis/crypto/config"
+	"io"
+)
+
+// This package provides  a dealer-less distributed verifiable secret sharing
+// using Pedersen VSS scheme as explained in "Provably Secure Distributed
+// Schnorr Signatures and a (t, n) Threshold Scheme for Implicit Certificates"
+// This file is only responsible for the setup of a shared secret among n peers.
+// The output is a global public polynomial (pubPoly) and a secret share for
+// each peers.
+
+// PolyInfo describe the information needed to construct (and verify) a matrixShare
+type PolyInfo struct {
+	// How many peer do we need to reconstruct a secret
+	T int
+	// How many peers do we need to verify
+	R int
+	// How many peers are collaborating into constructing the shared secret ( i.e. MatrixShare is of size NxN)
+	N int
+}
+
+// Represent the output of a VSS Pedersen scheme : a global public polynomial and a share of its related priv poly
+// for a peer
+type SharedSecret struct {
+
+	// The shared public polynomial
+	Pub *PubPoly
+
+	// The share of the shared secret
+	Share *abstract.Secret
+
+	// The index of this share regarding the secret private poly / pub poly
+	// i.e. it is the same as the receiver's index
+	Index int
+}
+
+// Dealer is a peer that will create a promise and distribute it to each receivers needed
+type Dealer struct {
```

Comment:
Yes it is absolutely workable. We should talk about wether we need the verification of the signature (validProof / blameProof) to happen when we generate the SharedSecret. If we do not **need** that, then a simple promise (so maybe renamed to Dealer) can do the trick and it is all good ! If we **do** need that verification (and if it is in promise, it must be for a good reason ;) then we need the dealer to send back its State to each peer after having received each signatures. I see two possible problems with that. **First**, the returned state would contain every shares of every one, and that is definitely not needed. **Secondly**, the terminology from the Receiver point of view would not be so elegant as we would have a :
* AddPromise (or AddDealer) (index,dealer)
* AddState (index, state) **MUST** be called at least T times ( for each dealer) before ProduceSharedSecret()

The first problem can be changed in a separate branch in the promise.go and do not impact this code. However, we must find a suitable "algorithm flow" for the second problem. Since it is closely related to how the promise works, we must talk in details about how are we going to restructure promise.go so this code can already be adapted to the future changes.